apollo-router/src/query_planner/caching_query_planner.rs

```diff
@@ -1,157 +1,165 @@
+                request.query_plan_options.clone(),
+            );
+            let context = request.context.clone();
+            let entry = qp.cache.get(&key).await;
+            return if entry.is_first() {
+                let res = qp.delegate.ready().await?.call(request).await;
+                match res {
+                    Ok(QueryPlannerResponse { content, context }) => {
+                        entry.insert(Ok(content.clone())).await;
+
+                        if let QueryPlannerContent::Plan { plan, .. } = &content {
+                            match (&plan.usage_reporting).serialize(Serializer) {
+                                Ok(v) => {
+                                    context.insert_json_value(USAGE_REPORTING, v);
+                                }
+                                Err(e) => {
+                                    tracing::error!(
+                                        "usage reporting was not serializable to context, {}",
+                                        e
+                                    );
+                                }
                             }
                         }
+                        Ok(QueryPlannerResponse { content, context })
                     }
-                    query_planner_content
-                })
-                .map_err(|e| {
-                    let e = e.into();
-
-                    let CacheResolverError::RetrievalError(re) = &e;
-                    if let QueryPlannerError::PlanningErrors(pe) = re.deref() {
-                        if let Err(inner_e) = request
-                            .context
-                            .insert(USAGE_REPORTING, pe.usage_reporting.clone())
-                        {
-                            tracing::error!(
-                                "usage reporting was not serializable to context, {}",
-                                inner_e
-                            );
-                        }
-                    } else if let QueryPlannerError::SpecError(e) = re.deref() {
-                        let error_key = match e {
-                            SpecError::ParsingError(_) => "## GraphQLParseFailure\n",
-                            _ => "## GraphQLValidationFailure\n",
-                        };
-                        if let Err(inner_e) = request.context.insert(
-                            USAGE_REPORTING,
-                            UsageReporting {
-                                stats_report_key: error_key.to_string(),
-                                referenced_fields_by_type: HashMap::new(),
-                            },
-                        ) {
-                            tracing::error!(
-                                "usage reporting was not serializable to context, {}",
-                                inner_e
-                            );
+                    Err(error) => {
+                        if let Some(error) = error.downcast_ref::<QueryPlannerError>() {
```

Comment:
Just a question, did you already test it ? I sometimes played with that and sometimes it doesn't work because you need to fetch the `error.source()`

Reply:
if a plugin replaces the error then that will not work, but there won't be anything we can do about it